3.2 - Changes required for #4283

lorenzo opened the pull request:
Moving result type conversion to the statements layer and renaming some functions that could cause confusion in the short term.

This is part of the work needed for automatically converting SQL functions to their corresponding PHP types. It also cleans up quite a bit the ResultSet class

lorenzo committed:
Replacing QueryExpression::type() with tieWith()

The old method name was confusing, and would become a barrier for
understanding how the new data type conversions for expressions
would work

lorenzo committed:
Moving result type conversion to the statements layer

This is part of the work needed for automatically converting SQL
functions to their correspondiny PHP types.

I also cleans up quite a bit the ResultSet class

dereuromark commented on src/Database/Expression/QueryExpression.php:
```diff
@@ -90,6 +90,19 @@ public function type($conjunction = null)
     /**
+     * Backwards compatible wrapper for tieWith()
+     *
+     * @param string $conjunction value to be used for joining conditions. If null it
```
string|null then for both methods

dereuromark commented on src/Database/Query.php:
```diff
@@ -1678,6 +1693,29 @@ public function bufferResults($enable = null)
+     *
+     * When called with no arguments, the current TypeMap object is returned.
+     *
+     * @param \Cake\Database\TypeMap $typeMap The map object to use
```
also |null

lorenzo committed:
Fixing doc blocks

lorenzo committed:
Re-applied an optimization that existed in the previous implementation.

String results will not be "casted" using the identity function found
in the StringType class, as a way of sparing some function calls.

lorenzo commented:
Just re-added an optimization that existed in the previous implementation. This time, using the `OptionalConverterInterface` instead of the old wacky inference.

markstory commented on src/Database/Dialect/SqlserverDialectTrait.php:
```diff
@@ -258,7 +258,7 @@ protected function _transformFunctionExpression(FunctionExpression $expression)
                 $expression
                     ->name('DATEADD')
-                    ->type(',')
+                    ->tieWith(',')
```
This is a nice change. `type(',')` was always a bit funny.

markstory commented on src/Database/FieldTypeConverter.php:
```diff
@@ -0,0 +1,82 @@
+        $this->_driver = $driver;
+        $map = $typeMap->toArray();
+        $types = array_keys(Type::map());
+        $types = array_map(['Cake\Database\Type', 'build'], array_combine($types, $types));
```
Should `Type` have a `buildAll()` method?

lorenzo replied:
sounds like a good idea

markstory commented on src/ORM/Query.php:
```diff
@@ -669,6 +670,8 @@ public function cleanCopy()
         $clone->offset(null);
         $clone->mapReduce(null, null, true);
         $clone->formatResults(null, true);
+        $clone->selectTypeMap(new TypeMap());
```
Should this be done in `__clone()` I don't know if the typemap should be shared between query clones.

lorenzo replied:
I can add that

lorenzo replied:
Added the clone. But I still need to completely reset the object here

markstory commented on src/ORM/Query.php:
```diff
@@ -920,6 +925,26 @@ protected function _addDefaultFields()
+    protected function _addDefaultSelectTypes()
+    {
+        $typeMap = $this->typeMap()->defaults();
+        $selectTypeMap = $this->selectTypeMap();
```
Why do the 2 type maps need be kept separate? For a select query, could we use only 1 TypeMap?

lorenzo replied:
Because they are not necessarily the same. More specifically, the select type map only contains the type for the aliases in the select clause. Sharing a single object for both would make the code quite difficult to manage.

A good example of not being the same is that you may want to add conditions using 'datetime' but at the end cast the column as a simple string.

markstory replied:
Very good point, I didn't consider that.

ADmad committed:
Fix CS error

ADmad commented:
I fixed a CS error :)

markstory commented on src/ORM/ResultSet.php:
```diff
@@ -602,14 +567,11 @@ protected function _groupResult($row)
     {
-        foreach ($this->_types[$alias] as $field => $type) {
-            $values[$field] = $type->toPHP($values[$field], $this->_driver);
-        }
```
This is nice. Having the caster attached as a decorator is great.

dereuromark commented on src/Database/Type/StringType.php:
```diff
@@ -94,4 +94,14 @@ public function marshal($value)
+     *
+     * @return boolean False as databse results are returned already as strings
+     */
+    public function requiresToPHPCast()
```
I think similar to https://github.com/cakephp/cakephp/blob/master/src/Database/Type/BoolType.php#L57 this should be called

    requiresToPhpCast()

(lowercase hp)

lorenzo committed:
Implemented a Type::buildAll() method to cleanup

This helped cleanup another part of the code

lorenzo committed:
Also cloning the select type map when cloning the query

lorenzo committed:
Changing casing of function in interface

lorenzo committed:
Merge branch '3.2-expression-data-type' of github.com:cakephp/cakephp into 3.2-expression-data-type

lorenzo commented:
@ADmad  Thanks
@markstory @dereuromark I took care of your feedback, thanks :)

lorenzo committed:
Changed the casing to another function

markstory commented:
Looks good to me :+1:
